### DISCUSSION TITLE
Implement a language server for Fastify

### DISCUSSION THREAD

**Discussion by mcollina** [fastify/fastify#2727] [Ideas]
Following this great article from @sw-yx on Language Servers https://www.swyx.io/language-servers/, I'm wondering if this would be something that could benefit our community.

**Comment by jsumners** [comment]
🤷‍♂️

**Comment by L2jLiga** [comment]
@mcollina what would you like to achive by language servers? I'm personally has no idea how fastify can benefit from it

**Reply by mcollina** [reply]
I was mainly inspired by the article. We could explore the idea of having a few eslint checks similar to https://reactjs.org/docs/hooks-rules.html#eslint-plugin to catch early problems.

**Reply by RafaelGSS** [reply]
For instance, when using `this` inside an arrow function handler, could throw a warning message.

**Reply by mcollina** [reply]
also using callbacks with `async` in hooks, passing `async` to `.listen()`

**Reply by simoneb** [reply]
> I was mainly inspired by the article. We could explore the idea of having a few eslint checks similar to https://reactjs.org/docs/hooks-rules.html#eslint-plugin to catch early problems.

I like the idea of having an eslint plugin for fastify. I find that signatures of functions and methods are sometimes not obvious and it's hard to say if you're using it properly or it's working just by chance. Two common examples of this are:

- when registering plugins by requiring the module which exports them inline, I often misplace parenthesis and instead of providing the plugin with options I'm providing a second argument to `require`, like so: 

```
fastify.register(require('some-module', { some: 'options' }))
```

- plugin signatures. for instance, fastify-cli complains when you export an async plugin with a single argument. I'm not sure why but it's something that eslint might catch
- handler signature: warnings about using callbacks/sending replies explicitly in async handlers or something like this

**Reply by StarpTech** [reply]
For me, a language server is fundamentally different than an eslint plugin to detect patterns. Do I miss something?

**Reply by simoneb** [reply]
At a very high level, my understanding of what eslint versus a language server extension do is: the former shows you errors after you've written code, the second gives you hints about writing code. Either way, both could be beneficial.

**Reply by voxpelli** [reply]
> At a very high level, my understanding of what eslint versus a language server extension do is: the former shows you errors after you've written code, the second gives you hints about writing code. Either way, both could be beneficial.

There's also intellisense that uses TypeScript types, so a language server essentially can fill any void ESLint and types (and probably shouldn't try to reimplement or duplicate something that ESLint or types can solve)

**Reply by StarpTech** [reply]
I think a good starting point would be to have a list of common pitfalls and bad patterns in Fastify. Regarding this, we can evaluate how to prevent them at an earlier stage.

**Reply by simoneb** [reply]
> > At a very high level, my understanding of what eslint versus a language server extension do is: the former shows you errors after you've written code, the second gives you hints about writing code. Either way, both could be beneficial.
> 
> There's also intellisense that uses TypeScript types, so a language server essentially can fill any void ESLint and types (and probably shouldn't try to reimplement or duplicate something that ESLint or types can solve)

We're looking into this in https://github.com/fastify/fastify/issues/2829, to provide intellisense where it can't be provided automatically.

**Reply by StarpTech** [reply]
I don't understand #2829. Why do you expect IntelliSense in pure JS? In order to provide IntelliSense you have to provide type information. That's why TS and TSDoc exists.

As far I understood the issue is about providing specific assistance in development with fastify (detect bad-patterns etc...).

**Reply by voxpelli** [reply]
> I don't understand #2829. Why do you expect IntelliSense in pure JS? In order to provide IntelliSense you have to provide type information. That's why TS and TSDoc exists.

Intellisense doesn't require TS or a 100% type coverage. When eg. importing a typed module much of the types can be inferred from the usage of that module, but when creating new functions one of course needs to annotate them to achieve intellisense and one can then chose to do so with JSDoc rather than migrate to TS.

> As far I understood the issue is about providing specific assistance in development with fastify (detect bad-patterns etc...).

Some anti-patterns could be detected through a plugin to a tool like ESLint, others through recommending better enforcement of types and, where possibly none of those two solve it, through some custom solution.

**Reply by StarpTech** [reply]
> Intellisense doesn't require TS or a 100% type coverage. When eg. importing a typed module much of the types can be inferred from the usage of that module, but when creating new functions one of course needs to annotate them to achieve intellisense and one can then chose to do so with JSDoc rather than migrate to TS.

That's correct, but again you can't expect anything from plain js and you need a way to annotate your code with additional information. #2829 just looks like a bad setup.

> Some anti-patterns could be detected through a plugin to a tool like ESLint, others through recommending better enforcement of types and, where possibly none of those two solve it, through some custom solution.

Agree

**Reply by voxpelli** [reply]
> you need a way to annotate your code with additional information

Yes? And JSDoc is that way?

**Comment by fox1t** [comment]
Seeing how the toolings in the JS world are moving on, I think developers would expect this kind of DX in the (near) future. 
Introducing a language server might also have another side benefit: fewer GitHub issues related to misunderstandings.

**Comment by simoneb** [comment]
Related issue https://github.com/fastify/fastify/issues/2829
